CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc

```diff
@@ -437,6 +451,42 @@ class LHCInfoPerFillPopConSourceHandler : public popcon::PopConSourceHandler<LHC
     return 0;
   }
 
+  cond::Time_t getFillLastLumiIOV(const cond::OMSService& oms, unsigned short fillId) {
+    // Define query to get the last run the fill
+    auto queryFills = oms.query("fills");
+    queryFills->addOutputVars({"last_run_number"});
+    queryFills->filterEQ("fill_number", fillId);
+
+    // Execute query
+    if (queryFills->execute()) {
+      auto queryFillsResult = queryFills->result();
+      if (!queryFillsResult.empty()) {
+        auto myLastRun = queryFillsResult.front().get<int>("last_run_number");
+
+        // Define the query to get the last lumi of the last run
+        auto queryRuns = oms.query("runs");
+        queryRuns->addOutputVars({"last_lumisection_number"});
+        queryRuns->filterEQ("run_number", myLastRun);
+
+        // Execute query
+        if (queryRuns->execute()) {
+          auto queryRunsResult = queryRuns->result();
+          if (!queryRunsResult.empty()) {
+            auto myLastLumi = queryRunsResult.front().get<unsigned short>("last_lumisection_number");
+            return cond::time::lumiTime(myLastRun, myLastLumi);
+          }
+        } else {
+          edm::LogInfo(m_name) << "OMS query for last_lumisection_number of run " << myLastRun
+                               << "failed, status:" << queryRuns->status();
+        }  // end last_lumi query
+      }
+    } else {
+      edm::LogInfo(m_name) << "OMS query for last_run_number of fill " << fillId
+                           << "failed, status:" << queryFills->status();
+    }  // end last_run query
+    return 1;
```

Comment:
I'm pretty sure this is not the correct value that should be returned in case the OMS queries fail, but I can't figure out which should be.... @JanChyczynski  any suggestion?

Reply:
If the query fails I don't think there's any sensible value we can return. I think we should just raise an exception here because it's not the responsibility of this function to decide what to do in such a case and what value should be used. 

Then there is a question of how to handle this exception and I don't really know but there are a couple of options:
In case of this exception, are we still able to insert the empty payload, i.e. obtain the IOV in a different way or calculate something like a default-fallback value? Maybe we could insert it on the lumi last processed by HLT (+ maybe a delay) - `getLastlumiProcessed`?
* In case we can: we empty the empty payload on the fallback IOV and log a working that the query failed and the empty payload may not be inserted on the right IOV
* If we can't insert the empty payload then we can either:
  *  Log an error about it and just continue
  * Not handle this exception and then it gets propagated upwards and probably it won't get caught and the PopCon execution ends with an error and hopefully someone sees that and can intervene

Reply:
> Maybe we could insert it on the lumi last processed by HLT (+ maybe a delay) - getLastlumiProcessed?

No I don't think that's feasible, unless we majorly restructure the whole O2O code.

> In case of this exception, are we still able to insert the empty payload, i.e. obtain the IOV in a different way or calculate something like a default-fallback value?

Ok then maybe my question is: if everything works ok we insert an empty payload with IOV = last lumi of the last run of the fill. Why do we need such an empty payload? Is there a "physics" reason?

Reply:
Summarising the private discussion with @JanChyczynski:
- the empty payload with IOV == "end of the fill" is needed to preserve the LHCInfo O2O logic
- when the OMS query fails, the best solution seems to throw a `cms::Exception`

I tried to add the throwing in the `else`s, but I'm stuck with the error:
> non-void function does not return a value in all control paths

Possible solutions:
- return something (not clear what)
- change the function itself to return `void` and use references to set the IOV
- ...?

@JanChyczynski any suggestion?

Reply:
If you put the `throw` statement just before the end of the function (i.e. where now it's `return 1`) the compiler is not going to complain.

Reply:
Thanks for the suggestion!
I followed it and also updated the query to the "lumisections" table, as we discussed privately.

This is addressed in 0d6f7c5e14a6556f630fd7618f9b21d6dbdd918a